### DISCUSSION TITLE
Async webserver

### DISCUSSION THREAD

**Discussion by lacssa** [Hieromon/AutoConnect#599] [General]
Hi Hieromon,

Great work you have been doing here!
I'm curious about any support for Async webservers.
I have been working on a GPS tracker which has a web interface, but has to be asynchronous since there are other processes in parallel.

Have you ever tried to check this out?

Cheers!

Best regards,
Luis Abreu

**Comment by Hieromon** [comment]
@lacssa Thank you for your interest in AutoConnect.

I am working on it right now. The AsyncAutoConnect library uses [ESPAsyncWebServer](https://github.com/me-no-dev/ESPAsyncWebServer) as its web server. Therefore, it can benefit from [AsyncTCP](https://github.com/me-no-dev/AsyncTCP) or [ESPAsyncTCP](https://github.com/me-no-dev/ESPAsyncTCP), which allows multiple connections. It also aims to ensure portability at the library API levels with ESP32 and ESP8266, similar to AutoConnect.

The current development status of AyncAutoConnect is in the process of implementing the basic AutoConnect functionality with the exception of custom web pages and the OTA feature.

Benefits that can be expected as:

1. Support for non-blocking mode when entering captive portal state with AsyncAutoConnect::begin.
2. Concurrent processing waiting for WiFi connection and WiFi scans. It works independently of the request handler for ESPAsyncWebServer, even when reconnecting in the background. Also there is no while loop.
3. Custom web pages are a callback from ESPAsyncWebServer that runs independently of the loop thread.

However, it is still possible that AsyncAutoConnect does not fully satisfy the following your mentions.
> has to be asynchronous since there are other processes in parallel.

The synchronization that is similar to AutoConnect will remain unresolved as:

4. _AsyncAutoConnect::handleClient_ is needed. DNS server presented by the core libraries is employed due to the unreliability of [Async  DNS server](https://github.com/devyte/ESPAsyncDNSServer) for spoofing captive portals. In other words, we need a role to drive _DNSServer::processNextRequest_ in a loop thread. We also need a driver for the AsyncAutoConnect service function that runs asynchronously. These are assumed to run in a loop thread, and the _AsyncAutoConnect::handleClient_ is responsible for that role.
5. You cannot use _yield_ or _delay_ or _while_ loops to create "waits" inside custom web page handlers. That probably causes WDT exceptions. Therefore, custom web pages will be limited to simple I/O. Instead AsyncAutoConnect supports the [Fetch API](https://hieromon.github.io/AutoConnect/acinteract.html#interact-with-sketches-by-autoconnectelements-event). It will be something like AutoConnect's [FetchLED.ino](https://github.com/Hieromon/AutoConnect/tree/master/examples/FetchLED) demo.

Do you have any specific wishes?

**Reply by lacssa** [reply]
Thanks for such a detailed answer. I have noticed similar issues with the DNS routing in the Async lib, great you also identified that. 

Handling DNS callbacks shouldn't be a big deal, we are aiming to optimize the loop time further to avoid congestion of requests.

Have you seen any issues with OTA + Async ? 

About the remaining updates, everything looks good. We will wait for the compete bundle (async + OTA). 

I will keep looking forward for any updates in this repo. 

Thanks !
